File: test_bowling.py
import pytest
from bowling import Game

@pytest.fixture
def game():
    return Game()

def roll_many(game, n, pins):
    for i in xrange(n):
        game.roll(pins)

def test_gutter_game(game):
    roll_many(game, 20, 0)

    assert game.get_score() == 0

def test_all_ones(game):
    roll_many(game, 20, 1)

    assert game.get_score() == 20

def test_one_spare(game):
    game.roll(5)
    game.roll(5)
    game.roll(3)
    roll_many(game, 17, 0)

    assert game.get_score() == 16

def test_false_spare(game):
    # If the scores that total to 10 are not in the same frame, it is not a spare.
    game.roll(0)
    game.roll(5)
    game.roll(5)
    game.roll(3)
    roll_many(game, 16, 0)

    assert game.get_score() == 13

def test_one_strike(game):
    game.roll(10)
    game.roll(3)
    game.roll(4)
    roll_many(game, 16, 0)

    assert game.get_score() == 24

def test_gutter_and_ten_means_spare(game):
    game.roll(0)
    game.roll(10)
    game.roll(3)
    game.roll(4)
    roll_many(game, 16, 0)

    assert game.get_score() == 20

def test_strike_followed_by_spare(game):
    game.roll(10)
    game.roll(3)
    game.roll(7)
    game.roll(4)
    game.roll(4)
    roll_many(game, 14, 0)

    assert game.get_score() == 42

def test_strike_followed_by_strike(game):
    game.roll(10)
    game.roll(10)
    game.roll(3)
    game.roll(4)
    roll_many(game, 14, 0)

    assert game.get_score() == 47

def test_spare_followed_by_strike(game):
    game.roll(3)
    game.roll(7)
    game.roll(10)
    game.roll(3)
    game.roll(4)
    roll_many(game, 14, 0)

    assert game.get_score() == 44

def test_last_frame_normal_case(game):
    roll_many(game, 18, 0)
    game.roll(3)
    game.roll(4)

    assert game.get_score() == 7

def test_last_frame_spare(game):
    roll_many(game, 18, 0)
    game.roll(3)
    game.roll(7)
    game.roll(9)

    assert game.get_score() == 19

@pytest.mark.skipif(True, reason='not implemented')
def test_last_frame_strike(game):
    roll_many(game, 18, 0)
    game.roll(10)
    game.roll(3)
    game.roll(4)

    assert game.get_score() == 17

@pytest.mark.skipif(True, reason='not implemented')
def test_last_frame_strike_followed_by_spare(game):
    roll_many(game, 18, 0)
    game.roll(10)
    game.roll(3)
    game.roll(7)

    assert game.get_score() == 20

@pytest.mark.skipif(True, reason='not implemented')
def test_perfect_game(game):
    roll_many(game, 12, 10)

    assert game.get_score() == 300

File: ruby/test_bowling.rb
require_relative './bowling.rb'
require 'test/unit'

class TestGame < Test::Unit::TestCase
  def setup
    @game = Game.new
  end

  def _roll_many(n, pins)
    (0..n).each do |i|
      @game.roll(pins)
    end
  end

  def test_gutter
    _roll_many(20, 0)
    assert_equal(@game.get_score(), 0)
  end

  def test_all_ones
    _roll_many(20, 1)
    assert_equal(@game.get_score(), 20)
  end

  def test_one_spare
    @game.roll(5)
    @game.roll(5)
    @game.roll(3)
    _roll_many(17, 0)

    assert_equal(@game.get_score(), 16)
  end

  def test_false_spare
    # If the scores that total to 10 are not in the same frame, it is not a spare.
    @game.roll(0)
    @game.roll(5)
    @game.roll(5)
    @game.roll(3)
    _roll_many(16, 0)
    assert_equal(@game.get_score(), 13)
  end

  def test_one_strike
    @game.roll(10)
    @game.roll(3)
    @game.roll(4)
    _roll_many(16, 0)

    assert_equal(@game.get_score(), 24)
  end

  def test_gutter_and_ten_means_spare
    @game.roll(0)
    @game.roll(10)
    @game.roll(3)
    @game.roll(4)
    _roll_many(16, 0)

    assert_equal(@game.get_score(), 20)
  end

  def test_strike_followed_by_spare
    @game.roll(10)
    @game.roll(3)
    @game.roll(7)
    @game.roll(4)
    @game.roll(4)
    _roll_many(14, 0)

    assert_equal(@game.get_score(), 42)
  end

  def test_strike_followed_by_strike
    @game.roll(10)
    @game.roll(10)
    @game.roll(3)
    @game.roll(4)
    _roll_many(14, 0)

    assert_equal(@game.get_score(), 47)
  end

  def test_spare_followed_by_strike
    @game.roll(3)
    @game.roll(7)
    @game.roll(10)
    @game.roll(3)
    @game.roll(4)
    _roll_many(14, 0)

    assert_equal(@game.get_score(), 44)
  end

  def test_last_frame_normal_case
    _roll_many(17, 0)
    @game.roll(3)
    @game.roll(4)

    assert_equal(@game.get_score(), 7)
  end

  def test_last_frame_spare
    _roll_many(18, 0)
    @game.roll(3)
    @game.roll(7)
    @game.roll(9)

    assert_equal(@game.get_score(), 19)
  end

  # The following tests are failing...
  #
  def test_last_frame_strike
    _roll_many(18, 0)
    @game.roll(10)
    @game.roll(3)
    @game.roll(4)

    assert_equal(@game.get_score(), 17)
  end

  def test_last_frame_strike_followed_by_spare
    _roll_many(18, 0)
    @game.roll(10)
    @game.roll(3)
    @game.roll(7)

    assert_equal(@game.get_score(), 20)
  end

  def test_perfect_game
    _roll_many(12, 10)

    assert_equal(@game.get_score(), 300)
  end

end
